core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -297,10 +418,16 @@ class LogSegment(val log: FileMessageSet,
   def lastModified = log.file.lastModified
 
   /**
+   * The largest timestamp this segment contains.
+   */
+  def largestTimestamp = if (maxTimestampSoFar >= 0) maxTimestampSoFar else lastModified
```

Comment:
Previously we don't have a concept of largest timestamp for a log segment. Implicitly the largest timestamp is the last modification time. So all the logic around timestamp is built based on last modification time. With KIP-33 supposedly most of the previous logic that are depending on ```lastModified()``` should now be using largest timestamp.

maxTimestampSoFar is only updated when it sees actual timestamp in a message (except for old inactive log segments where maxTimestampSoFar is always the same as last modification time and does not change).

I am thinking to leave ```lastModified()``` as it is but replace the external usage of ```lastModified()``` with ```largestTimestamp()```. ```largestTimestamp()``` prefers to use the maxTimestampSoFar if it exists, but falls back to lastModified if no message has a timestamp, such that the external logic would be the same for all segments no matter whether they contains timestamp or not. I may have missed some usages of ```lastModified()``` in the current patch. I will fix that in the coming patch.